[CORE-1576] Commands to set and unset primary domain.

mbaynton opened the pull request:
WIP: needs tests.

Should probably also add any reporting of current state onto this PR.

mbaynton committed:
Commands to set and unset primary domain.

WIP: needs tests.

greg-1-anderson commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,124 @@
+             */
+            list($site, $env) = $this->getSiteEnv($input->getArgument('site_env'));
+            $domains = $env->getDomains()->ids();
+            // Put pantheonsite.io domains at the bottom.
```
I think it would be better to prohibit setting `pantheonsite.io` as a primary domain, and just filter these out of the list.

mbaynton replied:
That's kind of a business rule implemented client-side. I agree it seems odd that you would set one of these domains as primary, but it's currently supported by the platform and if you can't do that, Terminus shouldn't be the only place that says no you can't do that. It's trivial to filter; can we remove it as a follow-up and talk to the rest of Core about whether to deny this?

greg-1-anderson replied:
The `getDomains` API returns all of the available domains, including the platform domains. We can't modify this API to not return platform domains, or we'd break other clients.

We could add an API `isPlatformDomain` or something like that. The two places that would need to use that API is this PR, and the Dashboard user interface for setting the primary domain, which hasn't been written yet.

The existing code already has explicit knowledge of what the platform domain is, in order to sort the platform domain to the end of the list. This logic, if maintained, would also need to be duplicated in the Dashboard, and could also benefit from an API to identify the platform domain (or even to sort the domain list).

I tend to think that these very specific APIs are not necessary, and that it would be okay to put this business logic in the two clients that need it. I think it's fine to raise the issue to `@core` and collect feedback from a wider audience. I think that filtering is better than changing the sort order, but I care 3/10.

mbaynton replied:
Oh, I didn't mean to change the list APIs, just that if we filter then I think the workflow ought to fail if asked to make a pantheon.io domain primary.

mbaynton replied:
I decided I was thinking with my developer hat, and not my UX hat. From the perspective of the "it is not easy to inadvertently set the wrong primary domain" perspective, filtering here but leaving it supported on the platform is exactly the right solution, as I discuss in most recent commit message.

greg-1-anderson commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,124 @@
+     *
+     * @authorize
+     *
+     * @command domain:primary:unset
```
Maybe `domain:primary:clear`?

Take a straw poll. 😄

mbaynton replied:
I realized that we're already calling this "reset"ting the primary domain in platform code and documentation; my calling it anything else is just a potential source of confusion. I've updated it to `domain:primary:reset` for consistency. How's that sound?

greg-1-anderson replied:
This is a product decision, ultimately. @ari-gold

greg-1-anderson replied:
("reset" sounds good to me)

ari-gold replied:
Let's go with "set" / "unset" ... I think "reset" is ambiguous and "unset" is more clear.

SaraMcCutcheon replied:
This needs to be `:remove` with an `:rm` alias to maintain consistency. Anything that is disassociated non-destructively (e.g. a domain doesn't disappear because you disassociate it from an environment) is `:remove`d, while anything that is disassociated destructively (e.g. removing a machine token destroys it) is `:delete`d.

ari-gold replied:
Thanks for the explanation @TeslaDethray.  I'm +1 on add/remove

P.S. It was helpful to search for various keywords (set, unset, add, remove, create, delete, enable, disable) on https://pantheon.io/docs/terminus/commands to see what our current conventions are.

mbaynton committed:
Tests for setting/resetting primary domain and its commands.

mbaynton committed:
Rename command to remove primary domain to `domain:primary:reset` for consistency.

mbaynton committed:
Refactor set_primary_domain workflow maker out of a Command to Models\Environment class.

mbaynton committed:
CS fix

mbaynton committed:
Filter pantheonsite.io domains from interactive choice list.

While the platform supports setting these domains as primary, it seems
highly unlikely that a user would do this intentionally, and more likely
that if it were chosen from a multiple choice list, that it was just
fat-fingered. For better UX, they are therefore excluded.

Users that really mean to do this can do so by typing/passing the full
pantheonsite.io domain non-interactively to the domain:primary:set
command.

mbaynton committed:
Add primary domain to Domain model and domain:list command.

greg-1-anderson commented on src/Commands/Domain/ListCommand.php:
```diff
@@ -29,6 +29,7 @@ class ListCommand extends TerminusCommand implements SiteAwareInterface
      * @field-labels
      *     id: Domain/ID
      *     type: Type
+     *     primary: Is Primary
```
I don't know if it's considered a b/c break to add a column to the default output. You could define `@default-fields: id, type, ... etc.` to maintain the same output. If you did that, though, then no one would see `primary` unless it was explicitly requested via `--fields`. I think it's probably better this way.

alexdicianu commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,104 @@
+
+        // The primary domain is set via a workflow so as to use workflow logging to track changes.
+        $this->log()->notice('Setting primary domain to {domain}...', ['domain' => $domain]);
+        $workflow = $env->setPrimaryDomain($domain);
```
Is there a possibility of verifying that the domain the user entered does actually belong to the current environment? We're already checking for this in the workflow but that's 1) not a good UX (start a workflow in the command line, check for potential errors on the dashboard) and 2) due to the way ygg handles workflow errors, the message itself is not always visible to the customer even on the dash (errors in tasks, for instance, are only visible in the debug tab).

alexdicianu replied:
I just pulled this on my local and it does show the workflow error message. Nvm :)

alexdicianu commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,104 @@
+        list(, $env) = $this->getSiteEnv($site_env);
+
+        $this->log()->notice('Unsetting primary domain...');
+        $workflow = $env->setPrimaryDomain(null);
```
same here

alexdicianu commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,104 @@
+             */
+            list($site, $env) = $this->getSiteEnv($input->getArgument('site_env'));
+            $domains = array_filter($env->getDomains()->ids(), function ($domain) {
+                return !preg_match('|\.pantheonsite\.io$|', $domain);
```
Is there a reason why we don't want to allow platform domains to be used as primary? I could imagine a situation where it can be useful during testing.

mbaynton replied:
Greg & I had extensive back and forth about this in this PR's conversation, the git log associated with the current state of things has the culmination of that. Maybe we could add it back into the list if the environment isn't live.

SaraMcCutcheon replied:
We'll want to extract this domain and move it somewhere, probably into the consts, so that it can be located and changed if necessary.

alexdicianu commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,104 @@
+        // The primary domain is set via a workflow so as to use workflow logging to track changes.
+        $this->log()->notice('Setting primary domain to {domain}...', ['domain' => $domain]);
+        $workflow = $env->setPrimaryDomain($domain);
+        $this->processWorkflow($workflow);
```
I think it would a good idea to print some sort confirmation message here (e.g. "primary domain was set to abc.com"). We show the progress bar but most often that is quite fast and at the end we're just left with `[notice] Setting primary domain to joytherapy.ro...`. I'm not sure what the status of the operation is at this point.

mbaynton replied:
Makes sense to me, this style just seemed to be what other workflow commands were doing :man_shrugging:

mbaynton committed:
UX review: Change command to domain:primary:unset, log completion.

SaraMcCutcheon commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,106 @@
+
+/**
+ * Class SetPrimaryCommand
+ * @package Pantheon\Terminus\Commands\Domain
```
This should be in `Pantheon\Terminus\Commands\Domain\Primary`

SaraMcCutcheon commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,106 @@
+     *
+     * @param string $site_env Site & environment in the format `site-name.env`
+     */
+    public function reset($site_env)
```
This function should be in a separate file called
`src/Commands/Domain/Primary/RemoveCommand` and the function should be called `remove`.

mbaynton replied:
I know the names are still being discussed and Naming is Hard(tm) but I went ahead with the s/unset/remove` half of it. Usability study of 1, I did feel distinct uncertainty whether I was instructing terminus to remove the domain that is primary from the environment, as opposed to removing the designation of primary, when I tested it.

SaraMcCutcheon commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,106 @@
+     *
+     * @usage domain:primary:set <site_env>
+     */
+    public function set($site_env, $domain)
```
For consistency, this command should be called `domain:primary:add`. Terminus' naming logic is that anything that is set cannot be unset but must always be set (e.g. a connection mode - you must have one) but things that can be removed are added (e.g. you can remove all the orgs from a site and it still works). Because this can be removed, it is an `:add` command.

ari-gold replied:
I'm not completely opposed to add/remove, but It think set/unset may be more clear and people may be wary to "remove" a primary domain thinking that would have the same effect as removing the domain completely from the environment (domain:remove). If it's remove...does removing the domain as "primary" also "remove" it from the site?

mbaynton replied:
I think the final decision is between @TeslaDethray and @ari-gold, but my 2c is that `add` also sounds more like an operation on a list of many. If I run (pseudocommands for brevity)
```
$ terminus domain:primary:add mydomain.com
$ terminus domain:primary:add www.mydomain.com
```
have I added two primary domains to my site? No.

ari-gold replied:
https://github.com/pantheon-systems/terminus/pull/2011#discussion_r326705988

SaraMcCutcheon commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,106 @@
+        list(, $env) = $this->getSiteEnv($site_env);
+
+        // The primary domain is set via a workflow so as to use workflow logging to track changes.
+        $this->log()->notice('Setting primary domain to {domain}...', ['domain' => $domain]);
```
A preemptive notice is not necessary. We only use them on commands which run more than one workflow.

SaraMcCutcheon commented on src/Commands/Domain/SetPrimaryCommand.php:
```diff
@@ -0,0 +1,106 @@
+        $this->log()->notice('Setting primary domain to {domain}...', ['domain' => $domain]);
+        $workflow = $env->setPrimaryDomain($domain);
+        $this->processWorkflow($workflow);
+        $this->log()->notice('Primary domain has been set to {domain}.', ['domain' => $domain]);
```
This notice should mention the site/env it was run on.

SaraMcCutcheon commented on src/Models/Environment.php:
```diff
@@ -650,6 +650,26 @@ public function importFiles($url)
+     *
+     * @return Workflow
+     */
+    public function setPrimaryDomain($domain)
```
This functionality should be in multiple functions so that its use is perfectly clear, preferably in a separate model belonging to Environment. You may want to take this functionality and create a new model, like `src/Models/Solr.php` does except you'll not extend `AddOnModel` but instead implement `EnvironmentInterface` and use `EnvironmentTrait`.

mbaynton replied:
> This functionality should be in multiple functions so that its use is perfectly clear

The function is currently 10 lines and 2 statements; the first statement sets local variable `$workflow` and the second returns it. It could be condensed to a single return statement. My point is just that I'm actually at a loss to think of how I can really break this up into more discrete functions/steps. Can you be more clear what you want with that?

mbaynton committed:
Changes that are clear from @TeslaDethray's requests.

mbaynton committed:
Rename domain:primary:set to domain:primary:add per consensus of product and technical leadership.

mbaynton committed:
Best guess at new model refactor request

mbaynton committed:
Fix failing tests in older minor versions of php 7.0 - 7.3.

It seems substr_compare issues a warning iff the first parameter is
the empty string in these earlier minor releases; manual states it
was fixed in 7.2.18. Since a) it's been fixed and b) empty strings
are not expected real-world input, I've simply removed that input
from the test coverage.

SaraMcCutcheon reviewed (changes requested):
This looks great! All that's needed is:
- Behat tests (which we've discussed already)
- CHANGELOG.md entries. This is what I'd write something like (and I'm quoting the help text here a bit):
```
### Added
- Added `primary` field to the output of `domain:list`. (#2011)
- Added `domain:primary:add` command to set a domain associated to the environment as primary. (#2011)
- Added `domain:primary:remove` command to remove a primary designation from a site environment. (#2011)
```

mbaynton committed:
Behat tests for add and remove primary domain

mbaynton committed:
Make behat.yml a bit more yaml-compliant (% not allowed as beginning of string)

mbaynton committed:
Restore Consolidation\\Filter\\Hooks\\FilterHooks to command list

mbaynton committed:
Changelog for 2.2.0

mbaynton commented:
The Scruitinizer build timed out after an hour failing to download things, and I don't see how to rerun it.
I took a guess at version # and date on the changelog.

SaraMcCutcheon commented on CHANGELOG.md:
```diff
@@ -1,6 +1,12 @@
 # Change Log
 All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org)
 
+## 2.2.0 - 2019-09-26
```
This line should say `## MASTER` until a release is tagged because this is going to be on `2.1.1-dev`.

mbaynton replied:
Well hey then we get to give Scrutinizer another go at it :)

mbaynton committed:
Fix stated version in changelog

SaraMcCutcheon reviewed (approved):
This is great! I think it can be merged as soon as you please.

garydylina reviewed (approved):
Checking whether my Approval will unblock Sara.  👍
